File: source/Samples/RevitMcpToolSet/Utilities/ParameterAccessor.cs
namespace RevitMcpToolSet.Utilities;

internal static class ParameterAccessor
{
    internal static string GetParameterValue(Parameter parameter)
    {
        try
        {
            return parameter.StorageType switch
            {
                StorageType.Integer => parameter.AsInteger().ToString(),
                StorageType.Double => parameter.AsDouble().ToString("G15"),
                StorageType.String => parameter.AsString() ?? "",
                StorageType.ElementId => parameter.AsElementId()?.ToString() ?? "",
                _ => parameter.AsValueString() ?? "",
            };
        }
        catch { return ""; }
    }

    internal static (bool success, string message, ElementId newElemId) ChangeType(Element element, long newTypeId)
    {
        try
        {
            var newId = element.ChangeTypeId(newTypeId.ToElementId());
            return (true, $"Change type to '{newTypeId}'", newId);
        }
        catch (Exception ex) when (ex is ArgumentException or InvalidOperationException)
        {
            return (false, $"Unexpected error changing type: {ex.Message}", ElementId.InvalidElementId);
        }
    }

    internal static (bool success, string message) SetParameterValue(Element element, string parameterName, string value)
    {
        try
        {
            var parameters = element.GetParameters(parameterName);
            if (parameters.Count == 0)
                return (false, $"Parameter '{parameterName}' not found");

            var writable = parameters.FirstOrDefault(p => !p.IsReadOnly);
            if (writable is null)
                return (false, $"Parameter '{parameterName}' is read-only");

            bool success;
            try
            {
                success = writable.StorageType switch
                {
                    StorageType.Integer when int.TryParse(value, out var intVal) => writable.Set(intVal),
                    StorageType.Integer => throw new FormatException($"Invalid integer value '{value}' for parameter '{parameterName}'"),
                    StorageType.Double when double.TryParse(value, out var dblVal) => writable.Set(dblVal),
                    StorageType.Double => throw new FormatException($"Invalid numeric value '{value}' for parameter '{parameterName}'"),
                    StorageType.String => writable.Set(value),
                    StorageType.ElementId when long.TryParse(value, out var eid) => writable.Set(eid.ToElementId()),
                    StorageType.ElementId => throw new FormatException($"Invalid element ID '{value}' for parameter '{parameterName}'"),
                    _ => throw new NotSupportedException($"Unsupported parameter storage type '{writable.StorageType}' for parameter '{parameterName}'"),
                };
            }
            catch (FormatException ex) { return (false, ex.Message); }
            catch (NotSupportedException ex) { return (false, ex.Message); }
            catch (InvalidOperationException ex) { return (false, $"Failed to set parameter '{parameterName}': {ex.Message}"); }

            return success
                ? (true, $"Set parameter '{parameterName}' to '{value}'")
                : (false, $"Parameter '{parameterName}' rejected the value '{value}'. The parameter may have constraints or the value may be out of range.");
        }
        catch (Exception ex)
        {
            return (false, $"Unexpected error setting parameter '{parameterName}': {ex.Message}");
        }
    }

    internal static string? GetBuiltInParam(Parameter parameter)
    {
        return (parameter.Definition as InternalDefinition)?.BuiltInParameter.ToString();
    }
}

File: samples/PythonDemo/mcp_toolset/shared/parameter_accessor.py
"""Parameter read/write helpers aligned with C# ParameterAccessor."""
from __future__ import annotations

from Autodesk.Revit import DB

from shared.element_helpers import element_id_value, normalize_string, param_value_as_string


def get_parameter_value(param: DB.Parameter, doc: DB.Document) -> str:
    if param is None or not param.HasValue:
        return ""
    return param_value_as_string(param, doc)


def set_parameter_value(element: DB.Element, param_name: str, value: str) -> tuple[bool, str]:
    param = element.LookupParameter(param_name)
    if param is None:
        return False, "Parameter '{}' not found".format(param_name)
    if param.IsReadOnly:
        return False, "Parameter '{}' is read-only".format(param_name)
    try:
        storage = param.StorageType
        if storage == DB.StorageType.String:
            param.Set(str(value))
        elif storage == DB.StorageType.Integer:
            param.Set(int(float(value)))
        elif storage == DB.StorageType.Double:
            param.Set(float(value))
        elif storage == DB.StorageType.ElementId:
            param.Set(DB.ElementId(int(value)))
        else:
            return False, "Unsupported storage type for '{}'".format(param_name)
        return True, ""
    except Exception as exc:
        return False, str(exc)


def change_element_type(element: DB.Element, new_type_id: int) -> tuple[bool, str]:
    try:
        element.ChangeTypeId(DB.ElementId(new_type_id))
        return True, ""
    except Exception as exc:
        return False, str(exc)


def parameter_entry(param: DB.Parameter, doc: DB.Document) -> dict:
    builtin = False
    try:
        definition = param.Definition
        if isinstance(definition, DB.InternalDefinition):
            builtin = definition.BuiltInParameter != DB.BuiltInParameter.INVALID
    except Exception:
        pass
    return {
        "name": normalize_string(param.Definition.Name),
        "value": get_parameter_value(param, doc),
        "storage": str(param.StorageType),
        "writable": not param.IsReadOnly,
        "builtin": builtin,
        "isShared": bool(param.IsShared),
        "hasValue": bool(param.HasValue),
    }
